File: USACO/2.3/concom.py
'''
ID: 11074501
TASK: concom
LANG: PYTHON3
'''


with open('concom.in', 'r') as f:
	n = int(f.readline())

	portfolio = {}
	owns = {}

	for a in range(n):
		thing = f.readline().split()
		i = int(thing[0])
		j = int(thing[1])
		p = int(thing[2])

		if i not in portfolio:
			portfolio[i] = {}

		portfolio[i][j] = p
		if portfolio[i][j] > 50:
			if i not in owns:
				owns[i] = []
			owns[i].append(j)


	for c, o in owns.items():
		bfs = [c]
		visited = {}

		all_owns = []
		while(len(bfs) > 0):
			consider = bfs.pop()
			if consider in visited:
				continue
			visited[consider] = 1

			if consider in owns:
				for children in owns[consider]:
					bfs.append(children)
					if children not in all_owns:
						all_owns.append(children)
		owns[c] = all_owns

	for company, port in portfolio.items():
		updates = {}

		if company in owns:

			for subseq in owns[company]:

				if subseq not in portfolio:
					continue

				for new_company, new_shares in portfolio[subseq].items():

					if new_company not in port:
						port[new_company] = 0
					port[new_company] += new_shares

					if port[new_company] > 50 and new_company not in owns[company]:
						owns[company].append(new_company)

	ans = []
	for company, port in portfolio.items():
		for k, v in port.items():
			if v > 50:
				if k!=company:
					ans.append((company, k))

	ans.sort()

	with open('concom.out', 'w') as fo:
		for elem in ans:
			fo.write(str(elem[0]) + ' ' + str(elem[1]) + '\n')

File: USACO/2.3/concom.cpp
#include <bits/stdc++.h>
using namespace std;
#define loop(i, a) for (int i = 0; i < a; i++)

int portfolio[100][100];

/*
ID: 11074501
TASK: concom
LANG: C++
*/


int main(){
	ofstream fout ("concom.out");
	ifstream fin ("concom.in");

	int n;
	fin >> n;

	memset(portfolio, 0, sizeof portfolio);
	map<int, vector<int>>owns;

	while(n--){
		int i, j, p;
		fin >> i >> j >> p;
		portfolio[i][j] = p;
		if (p > 50) owns[i].push_back(j);
	}

	// bfs to update owns
	for (int i = 0; i < owns.size(); i++){
		queue<int>bfs;
		bfs.push(i);

		while(!bfs.empty()){
			int thing = bfs.front();
			bfs.pop();

			vector<int> children = owns[thing];

			for (int j = 0; j < children.size(); j++){
				bfs.push(children[j]);
				if (find(owns[i].begin(), owns[i].end(), children[j]) == owns[i].end()){
					owns[i].push_back(children[j]);
				}
			}
		}
		
		vector<int>updates;
		for (int j = 0; j < owns[i].size(); j++){
			int new_company = owns[i][j];
			for (int k = 0; k < 100; k++){
				portfolio[i][k] += portfolio[new_company][k];
				
				if (portfolio[i][k] > 50 && 
				find(owns[i].begin(), owns[i].end(), k) == owns[i].end() && 
				find(updates.begin(), updates.end(), k) == updates.end()){
					updates.push_back(k);
				}
			}
		}
		
		for (int j = 0; j < updates.size(); j++) owns[i].push_back(updates[j]);
	}
	
	bool printed[100][100];
	memset(printed, false, sizeof printed);
	
	for (int i = 0; i < owns.size(); i++){
		sort (owns[i].begin(), owns[i].end());
		for (int j = 0; j < owns[i].size(); j++){
			if (printed[i][j] == false && printed[j][i] == false){
				fout << i << " " << owns[i][j] << endl;
			}
			printed[i][j] = true;
			printed[j][i] = true;
		}
	}
}